### PR TITLE
[SPARK-16678] [SPARK-16677] [SQL] Fix two View-related bugs

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?

**Issue 1: Disallow Creating/Altering a View when the same-name Table Exists (without IF NOT EXISTS)**
When we create OR alter a view, we check whether the view already exists. In the current implementation, if a table with the same name exists, we treat it as a view. However, this is not the right behavior. We should follow what Hive does. For example,

```
hive> CREATE TABLE tab1 (id int);
OK
Time taken: 0.196 seconds
hive> CREATE OR REPLACE VIEW tab1 AS SELECT * FROM t1;
FAILED: SemanticException [Error 10218]: Existing table is not a view
 The following is an existing table, not a view: default.tab1
hive> ALTER VIEW tab1 AS SELECT * FROM t1;
FAILED: SemanticException [Error 10218]: Existing table is not a view
 The following is an existing table, not a view: default.tab1
hive> CREATE VIEW IF NOT EXISTS tab1 AS SELECT * FROM t1;
OK
Time taken: 0.678 seconds
```

**Issue 2: Strange Error when Issuing Load Table Against A View**
Users should not be allowed to issue LOAD DATA against a view. Currently, when users doing it, we got a very strange runtime error. For example,

``` SQL
LOAD DATA LOCAL INPATH "$testData" INTO TABLE $viewName
```

```
java.lang.reflect.InvocationTargetException was thrown.
java.lang.reflect.InvocationTargetException
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:606)
    at org.apache.spark.sql.hive.client.Shim_v0_14.loadTable(HiveShim.scala:680)
```
## How was this patch tested?

Added test cases

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala [hunk 1]
```diff
@@ -105,6 +105,12 @@ case class CreateViewCommand(
       val tableIdentifier = tableDesc.identifier.copy(database = Option(database))
 
       if (sessionState.catalog.tableExists(tableIdentifier)) {
+        val tableMetadata = sessionState.catalog.getTableMetadata(tableIdentifier)
+        if (tableMetadata.tableType != CatalogTableType.VIEW) {
+          throw new AnalysisException(
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: If allowing existing, do we still need to throw this excepion?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala]: Yeah, a good question. Just did a check using Hive. It is allowed by Hive. 
```
hive> CREATE VIEW IF NOT EXISTS tab1 AS SELECT * FROM t1;
OK
Time taken: 0.678 seconds
```
Let me move this a little bit down.

**Comment by gatorsmile** [comment]
cc @cloud-fan

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 2]
```diff
@@ -55,6 +54,76 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
 
+  test("error handling: existing a table with the duplicate name when creating/altering a view") {
+    withTable("tab1") {
+      sql("CREATE TABLE tab1 (id int)")
+      var e = intercept[AnalysisException] {
+        sql("CREATE OR REPLACE VIEW tab1 AS SELECT * FROM jt")
+      }.getMessage
+      assert(e.contains("The following is an existing table, not a view: `default`.`tab1`"))
+      e = intercept[AnalysisException] {
+        sql("CREATE VIEW tab1 AS SELECT * FROM jt")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: what's hive's error message for this case? `view already existing` or `not a view`?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala [hunk 3]
```diff
@@ -55,6 +54,76 @@ class SQLViewSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+    withTable("tab1") {
+      sql("CREATE TABLE tab1 (id int)")
+      sql("CREATE VIEW IF NOT EXISTS tab1 AS SELECT * FROM jt")
+      checkAnswer(sql("select count(*) FROM tab1"), Row(0))
+    }
+  }
+
+  test("error handling: insert/load/truncate table commands against a temp view") {
+    val viewName = "testView"
+    withView(viewName) {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: `withTempTable`, as now we call temp view as temp table in test utils...

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: For this case, it shows `table already existed`. However, when we found a view/table exists, the error message is wrong:
```
View $tableIdentifier already exists. If you want to update the view definition,
 please use ALTER VIEW AS or CREATE OR REPLACE VIEW AS
```
We are unable to alter a table (that is not a view) by these commands `ALTER VIEW AS or CREATE OR REPLACE VIEW AS`.

See the [source code] (https://github.com/gatorsmile/spark/blob/c64092c6aaec42663278343a27467c1c8c165b92/sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala#L119-L124)

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]: Will fix it, after your PR is merged. https://github.com/apache/spark/pull/14318

Thanks!

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLViewSuite.scala]:
```
hive> CREATE VIEW tab1 AS SELECT * FROM t1;
FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.DDLTask. AlreadyExistsException(message:Table tab1 already exists)
```

**Comment by cloud-fan** [comment]
LGTM, waiting for https://github.com/apache/spark/pull/14297

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**